sklearn/utils/estimator_checks.py

```diff
@@ -3193,7 +3255,25 @@ def check_n_features_in_after_fitting(name, estimator_orig, strict_mode=True):
         estimator.partial_fit(X_bad, y)
 
 
-# set of checks that are completely strict, i.e. they have no non-strict part
-_FULLY_STRICT_CHECKS = set([
-    'check_n_features_in',
+# set of checks that do not check API-compatibility. They are ignored when
+# api_only is True.
+_NON_API_CHECKS = set([
```

Comment:
I don't think we can immediately do that, because there are 3 kinds of checks, not 2:

1. the ones that don't check the API at all (in this dict)
2. the ones that only check the API
3. the ones that check the API in some parts, and do some other kinds of checks in other parts.

1 and 2 don't use the `api_only` parameter. The checks in 3 are the only ones to use it (we agreed in previous PRs that this was the easiest way to go).

separating 2 and 3 would be a lot of work, probably worth it I agree, but out of scope here.

Reply:
> separating 2 and 3 would be a lot of work, probably worth it I agree, but out of scope here.

I see defining the API mainly about communication. Having 3 makes it harder to communicate what the API is.

I guess as a first step, we can add this `api_only=True`, and state "Your estimator is API compliant if you pass `check_estimator(..., api_only=True)`". In a future PR, we can separate 2 and 3 and then we can firmly say "These are the checks that test the API".

Reply:
Agreed. Merging this PR isn't the end of the road, it's only the beginning.
Which is why I think we should keep things as simple as possible here, and improve incrementally in the future.